### PR TITLE
fixes 2 alter partitioned table bugs

### PR DESCRIPTION
**Pull request by seut**
- alter empty partitioned table settings changes
   setting on all tables
 - changing the column policy on a partitioned table
   throws an exception if more than 1 partition exists.

### CODE REVIEW

File: sql/src/test/java/io/crate/integrationtests/ColumnPolicyIntegrationTest.java [hunk 1]
```diff
@@ -567,6 +567,7 @@ public void testAlterPartitionedTable() throws Exception {
                 ") partitioned by (score) with (number_of_replicas=0, column_policy='strict')");
         ensureYellow();
         execute("insert into dynamic_table (id, score) values (1, 10)");
+        execute("insert into dynamic_table (id, score) values (1, 20)");
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/ColumnPolicyIntegrationTest.java]: why is this change necessary?

Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/ColumnPolicyIntegrationTest.java]: this is the 2nd bug mentioned at the commit message

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 2]
```diff
@@ -125,50 +126,50 @@ public void onFailure(Throwable e) {
...
-                // template gets all changes unfiltered
-                results.add(updateTemplate(analysis.tableParameter(), table.ident()));
-
                 if (!analysis.excludePartitions()) {
-                    // resolve indices on master node to make sure it doesn't hit partitions that are being deleted
-                    String index = table.ident().indexName();
-                    results.add(updateMapping(analysis.tableParameter().mappings(), index));
-                    results.add(updateSettings(parameterWithFilteredSettings, index));
+                    operations.set(3);
+                    String[] indices = getIndexNames(table, null);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: why not just call table.concreteIndices? you're already in if blocks that are repeated within getIndexNames

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 3]
```diff
@@ -114,7 +113,9 @@ public void onFailure(Throwable e) {
             return Futures.immediateFailedFuture(new AlterTableAliasException(table.ident().fqn()));
         }
 
-        List<ListenableFuture<Long>> results = new ArrayList<>(3);
+        final SettableFuture<Long> result = SettableFuture.create();
+        final AtomicInteger operations = new AtomicInteger(2);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: why change everything to the operations pattern instead of just adding the result of updateMapping/updateSettings to the results list where it was missing?

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 4]
```diff
@@ -330,21 +301,41 @@ public void onFailure(Throwable e) {
...
+        public void onResponse(T putMappingResponse) {
+            if (operations.decrementAndGet() == 0) {
+                result.set(1L);
+            }
         }
 
         @Override
         public void onFailure(Throwable e) {
-            future.setException(e);
+            result.setException(e);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: what happens to other pending operations in an error case?

File: sql/src/test/java/io/crate/integrationtests/ColumnPolicyIntegrationTest.java (hunk 1 above)
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/ColumnPolicyIntegrationTest.java]: can you change the method name to clarify what is being tested here?

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java (hunk 3 above)
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: where is the benefit of the other pattern?

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java (hunk 4 above)
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: same like before, they can't be stopped so they will keep running but I don't see a need to wait for them

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java (hunk 3 above)
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: It avoids having to adjust the number of operations and also doesn't require `if (operations.decrementAndGet() == 0) {` logic at various places.

Imho it is easier to extend, but maybe I'm missing something

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 5]
```diff
@@ -123,52 +124,47 @@ public void onFailure(Throwable e) {
...
+    private ListenableFuture<Long> updateTemplate(Map<String, Object> newMappings,
+                                                  Settings newSettings,
+                                                  TableIdent tableIdent) {
         String templateName = PartitionName.templateName(tableIdent.schema(), tableIdent.name());
         IndexTemplateMetaData indexTemplateMetaData =
                 clusterService.state().metaData().templates().get(templateName);
         if (indexTemplateMetaData == null) {
-            templateFuture.setException(new RuntimeException("Template for partitioned table is missing"));
-            return;
+            Futures.immediateFailedFuture(new RuntimeException("Template for partitioned table is missing"));
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: return is missing?

File: sql/src/test/java/io/crate/integrationtests/ColumnPolicyIntegrationTest.java [hunk 6]
```diff
@@ -560,21 +560,23 @@ public void testResetColumnPolicyPartitioned() throws Exception {
...
 
     @Test
-    public void testAlterPartitionedTable() throws Exception {
+    public void testAlterColumnPolicyOnPartitionedTableWithExistingPartitions() throws Exception {
         execute("create table dynamic_table (" +
                 "  id integer, " +
                 "  score double" +
                 ") partitioned by (score) with (number_of_replicas=0, column_policy='strict')");
         ensureYellow();
+        // create at least 2 partitions to test real multi partition logic (1 partition would behave similar to 1 normal table)
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/ColumnPolicyIntegrationTest.java]: :+1:

**Comment by mfussenegger** [comment]
ack
